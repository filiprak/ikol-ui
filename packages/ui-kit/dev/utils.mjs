import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
