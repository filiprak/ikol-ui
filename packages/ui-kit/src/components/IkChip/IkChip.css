.ik-chip {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);

    border-radius: 50px;
    height: 32px;
    line-height: 24px;
    max-width: 100%;

    display: inline-flex;
}

.ik-chip--disabled {
    opacity: 0.85;
    pointer-events: none;
}

.ik-chip--primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-chip--success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-chip--error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-chip--accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-chip__content:not(:last-child) {
    padding-right: 2px;
}

.ik-chip__actions {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    align-content: center;
    padding: 0 6px 0 4px;
}

.ik-chip--size-sm .ik-chip__actions {
    padding: 0 2px 0 3px;
}

.ik-chip--size-sm {
    font-size: 12px;
    font-weight: 400;
    height: 20px;
}

.ik-chip__content {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    align-content: center;
}

.ik-chip--size-sm .ik-chip__actions .ik-icon {
    font-size: 10px !important;
}

.ik-chip--size-md .ik-chip__actions .ik-icon {
    font-size: 12px !important;
}