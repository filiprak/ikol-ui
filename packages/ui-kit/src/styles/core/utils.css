/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
