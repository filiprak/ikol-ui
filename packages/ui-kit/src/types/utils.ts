import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
