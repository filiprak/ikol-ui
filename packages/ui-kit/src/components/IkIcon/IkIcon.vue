<script setup lang="ts">
import '@ui/styles';
import './IkIcon.css';
import { type PropType, h, useAttrs } from 'vue';
import { toCssUnits, getIconClasses } from '@ui/utils/helpers';
import { useRender } from '@ui/composables/render';
import type { IkUIDesignColor } from '@ui/types/utils';

const props = defineProps({
    tag: {
        type: String,
        default: 'i',
    },
    icon: {
        type: String,
    },
    size: {
        type: [Number, String],
    },
    size_px: {
        type: [Number, String],
    },
    disabled: {
        type: Boolean,
        default: false,
    },
    circle: {
        type: Boolean,
        default: false,
    },
    design: {
        type: String as PropType<`${IkUIDesignColor}`>,
    },
    variant: {
        type: String,
    },
});
const attrs = useAttrs();

useRender(() => {
    return h(props.tag, {
        class: [
            {
                'ik-icon': true,
                ['ik-icon--' + props.design]: !!props.design,
                ['ik-icon--' + props.variant]: !!props.variant,
                'ik-icon--circle': props.circle,
                'ik-icon--clickable': !props.disabled && !!attrs.onClick,
                'ik-icon--disabled': props.disabled,
                ['fa-' + props.size + 'x']: !!props.size,
            },
            getIconClasses(props.icon || ''),
        ],
        style: {
            width: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            height: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            fontSize: props.size_px ? toCssUnits(props.circle ? (Number(props.size_px) / 2) : props.size_px, 'px') : null,
        },
    });
});
</script>
