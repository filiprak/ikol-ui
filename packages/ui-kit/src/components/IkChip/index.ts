import IkChip from './IkChip.vue';
export interface IkChip extends InstanceType<typeof IkChip> { };
export { IkChip };
