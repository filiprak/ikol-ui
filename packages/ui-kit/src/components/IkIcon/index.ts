import IkIcon from './IkIcon.vue';
export interface IkIcon extends InstanceType<typeof IkIcon> { };
export { IkIcon };
