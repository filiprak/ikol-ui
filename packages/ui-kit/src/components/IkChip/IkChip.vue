<template>
    <div :class="root_class"
         :style="chip_style">
        <div class="ik-chip__content"
             :style="chip_content_style">
            <slot></slot>
        </div>
        <div v-if="removable || custom_action"
             class="ik-chip__actions">
            <IkIcon v-if="custom_action"
                    class="ik-chip__action"
                    :size_px="icon_size"
                    :icon="custom_action"
                    :style="custom_icon_style"
                    circle
                    @click.stop="$emit('ik-custom-click')"
                    @mousedown.stop
                    @mouseenter="custom_hover = true"
                    @mouseleave="custom_hover = false" />
            <IkIcon v-if="removable"
                    class="ik-chip__action"
                    :size_px="icon_size"
                    icon="times:light"
                    :style="remove_icon_style"
                    circle
                    @click.stop="$emit('ik-remove')"
                    @mousedown.stop
                    @mouseenter="remove_hover = true"
                    @mouseleave="remove_hover = false" />
        </div>
    </div>
</template>
<script setup lang="ts">
import '@ui/styles';
import './IkChip.css';
import { IkIcon } from '@ui/components/IkIcon';
import { ref, computed, type PropType } from 'vue';
import { useTheme } from '@ui/composables/theme';
import { getTextColorForBackground } from '@ui/utils/helpers';
import type { IkUIDesignColor } from '@ui/types/utils';

const custom_hover = ref(false);
const remove_hover = ref(false);

const theme = useTheme();

const design_colors: Record<`${IkUIDesignColor}`, string | null> = {
    primary: theme.colors.value.primary_color,
    accent: theme.colors.value.accent_color,
    error: theme.colors.value.error_color,
    success: theme.colors.value.success_color,
    default: theme.colors.value.on_surface_color_3,
};

const props = defineProps({
    /**
    * Color theme variant. Possible values are `"accent", "error", "success", "primary", "default"`.
    * @instance
    * @type String
    * @default undefined
    */
    design: {
        type: String as PropType<`${IkUIDesignColor}`>,
    },
    /**
     * Whether to append remove button at the end of chip content.
     * @instance
     * @type Boolean
     * @default false
     */
    removable: {
        type: Boolean,
        default: false,
    },
    /**
     * Whether component should not accept mouse events.
     * @instance
     * @type Boolean
     * @default false
     */
    disabled: {
        type: Boolean,
        default: false,
    },
    /**
     * Adds custom action to chip, string represents icon name that will be displayed.
     * @instance
     * @type String
     * @default undefined
     */
    custom_action: {
        type: String,
    },
    /**
     * Chip size. Possible values are: "sm", "md".
     * @instance
     * @type String
     * @default 'md'
     */
    size: {
        type: String,
        default: 'md',
    },
    /**
     * Chip backgorund color.
     * @instance
     * @type String
     * @default undefined
     */
    color: {
        type: String,
    },
});

if (props.design && props.color) {
    // eslint-disable-next-line no-console
    console.error('Props design and color should not be used together');
}

function _getIconStyle(is_hovering: boolean) {
    const style = {
        backgroundColor: '',
        color: '',
    };
    if (props.design && props.design !== 'default') {
        const design_color = '#' + design_colors[props.design];

        if (is_hovering) {
            style.backgroundColor = '#' + getTextColorForBackground(design_color);
            style.color = design_color;
        } else {
            style.backgroundColor = design_color;
            style.color = '#' + getTextColorForBackground(design_color);
        }
    } else {
        if (is_hovering) {
            style.backgroundColor = props.color ? '#' + getTextColorForBackground(props.color) : 'var(--on-surface-darken-color-3)';
            style.color = props.color ? '#' + props.color : 'var(--surface-darken-color)';
        } else {
            style.backgroundColor = props.color ? '#' + props.color : 'var(--surface-darken-color)';
            style.color = props.color ? '#' + getTextColorForBackground(props.color) : 'var(--on-surface-darken-color-3)';
        }
    }

    return style;
};

const root_class = computed(() => {
    const classes = ['ik-chip'];

    if (props.removable) {
        classes.push('ik-chip--removable');
    }
    if (props.disabled) {
        classes.push('ik-chip--disabled');
    }
    if (props.design) {
        classes.push('ik-chip--' + props.design);
    }
    if (props.size) {
        classes.push('ik-chip--size-' + props.size);
    }

    return classes;
});

const chip_style = computed(() => {
    return props.color
        ? {
            backgroundColor: '#' + props.color,
            color: '#' + getTextColorForBackground(props.color),
        }
        : {};
});

const custom_icon_style = computed(() => {
    return _getIconStyle(custom_hover.value);
});

const remove_icon_style = computed(() => {
    return _getIconStyle(remove_hover.value);
});

const chip_content_style = computed(() => {
    if (props.size === 'sm') {
        return {
            padding: props.removable || props.custom_action ? '4px 0 4px 8px' : '4px 8px',
        };
    } else {
        return {
            padding: props.removable || props.custom_action ? '4px 0 4px 16px' : '4px 16px',
        };
    }
});

const icon_size = computed(() => {
    return props.size == 'sm' ? 16 : 20;
});

</script>
