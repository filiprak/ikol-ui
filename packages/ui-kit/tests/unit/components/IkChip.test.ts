import { renderDeep } from '@ui/tests/tools/utils';
import { IkChip } from '@ui/components/IkChip';
import { IkIcon } from '@ui/components/IkIcon';

describe('IkChip', () => {
    const getInstance = (options = {}) => renderDeep(IkChip, options);

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toContain('ik-chip');
        expect(wrapper.classes()).toContain('ik-chip--size-md');
        expect(wrapper.element.tagName).toBe('DIV');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders disabled', () => {
        const wrapper = getInstance({
            props: {
                disabled: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-chip--disabled');
    });

    it('renders with default slot text', () => {
        const wrapper = getInstance({
            slots: {
                default: 'Test',
            },
        });
        expect(wrapper.text()).toContain('Test');
    });

    it('renders with design parameter', () => {
        const wrapper = getInstance({
            props: {
                design: 'primary',
            },
        });
        expect(wrapper.classes()).toContain('ik-chip--primary');
    });

    it('renders removable', () => {
        const wrapper = getInstance({
            props: {
                removable: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-chip--removable');
        expect(wrapper.findComponent(IkIcon).props('icon')).toContain('times:light');
    });

    it('renders with custom_action', () => {
        const wrapper = getInstance({
            props: {
                custom_action: 'pen',
            },
        });
        expect(wrapper.findComponent(IkIcon).props('icon')).toContain('pen');
    });

    it('renders with size', () => {
        const wrapper = getInstance({
            props: {
                size: 'sm',
            },
        });
        expect(wrapper.classes()).toContain('ik-chip--size-sm');
    });

    it('renders with custom background color', () => {
        const wrapper = getInstance({
            props: {
                color: 'ffa500',
            },
        });
        expect(wrapper.html()).toContain('background-color: rgb(255, 165, 0);');
    });
});
