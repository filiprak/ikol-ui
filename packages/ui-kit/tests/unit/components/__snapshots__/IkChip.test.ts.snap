// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkChip renders with correct default state 1`] = `
<div
  class="ik-chip ik-chip--size-md"
>
  <div
    class="ik-chip__content"
    style="padding: 4px 16px;"
  >
    
    
  </div>
  <!--v-if-->
</div>
`;
